Support non null constraints - JDBC support #347

amitvc opened the pull request:
… supported type

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Rember to add the correct license header to new files.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

amitvc commented:
@eolivelli I added a UT in jdbc/SystemTablesTest.java to demonstrate that we throw the exception when not null constraint is applied for non supported types. Appropriate exception is thrown here - https://github.com/diennea/herddb/blob/master/herddb-core/src/main/java/herddb/server/ServerSideConnectionPeer.java#L705

Currently we dont catch this exception and hence it is not being propagated back to the client over the network. I am guessing we need to handle these things and also interpret them on the client side. Do we have existing protocol or message types defined that we use to communicate these things ?

Exception --
herddb.model.StatementExecutionException: Not null constraints not supported for column type 6
at herddb.model.ColumnTypes.getNonNullTypeForPrimitiveType(ColumnTypes.java:105)
at herddb.sql.DDLSQLPlanner.buildCreateTableStatement(DDLSQLPlanner.java:369)
at herddb.sql.DDLSQLPlanner.plan(DDLSQLPlanner.java:293)
at herddb.sql.DDLSQLPlanner.translate(DDLSQLPlanner.java:270)
at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:242)
at herddb.server.ServerSideConnectionPeer.handleExecuteStatement(ServerSideConnectionPeer.java:705)

eolivelli commented:
You are right.
Good catch
It is a regression introduced by the refactoring in which we introduced CompletableFuture in this class.

Please add a try/catch.
In the sql client you should expect a SQLException.

Please check other usages of 'translate' in this class and add tests for them.

amitvc commented:
@eolivelli - Added more tests for different data types. But these tests are all about table creation. 
Are there already tests for data insertion or updates in the jdbc module where we can add tests for not null constraints on types that are supported and not supported?

eolivelli commented on herddb-jdbc/src/test/java/herddb/jdbc/SystemTablesTest.java:
```diff
@@ -227,4 +225,148 @@ public void test() throws Exception {
+                try (BasicHerdDBDataSource dataSource = new BasicHerdDBDataSource(client);
+                     Connection con = dataSource.getConnection();
+                     Statement statement = con.createStatement();) {
+                     statement.execute("CREATE TABLE mytable (key string primary key, name string,b1 boolean not null)");
```
please add a test case with PreparedStatement#addBatch and then executeBatch, it will run thru a different code path

eolivelli commented on herddb-jdbc/src/test/java/herddb/jdbc/SystemTablesTest.java:
```diff
@@ -227,4 +225,148 @@ public void test() throws Exception {
+    }
+
+    @Test(expected = SQLException.class)
+    public void ensureWeThrowExceptionForNonSupportedNotNullTypes_BOOLEAN() throws Exception {
```
we can keep just this one and a test which "works", there is no much value in testing other datatype

eolivelli commented:
We have to add a test about DatabaseMetadata#getColumns as well, then we are done with this new feature!

thank you very much @amitvc

amitvc commented:
@eolivelli - Addressed comments and added tests.
Regarding DatabaseMetadata#getColumns I have added some unit test for non null types. I think ColumnTypes.typeToString method for not null types is return text like "{type} not null". In the unit test for the DatabaseMetadata.getColumns i had to assert against that types. I feel that is not correct. If the type is String or NOTNULL_STRING it should still return 'string' as a type. What do you think?

eolivelli commented:
@amitvc 
these are the specs of getColumns:

https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getColumns(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String)

column 5 is
DATA_TYPE  int => SQL type from java.sql.Types 
-> we have to follow the standard java.sql.Types 

column N6  is:
TYPE_NAME String => Data source dependent type name, for a UDT the type name is fully qualified 
-> so we are free to return whatever we want, having 'string not null' would be fine, but I feel it is better to return just "string" 

column N18 is:
IS_NULLABLE String => ISO rules are used to determine the nullability for a column. 
I am expecting here that the result reflect your change

amitvc commented:
@eolivelli I will address 'string not null' in a separate task since this PR is closed. Please assign me more work or tickets. I am going to play around with herddb with a cluster setup over the weekend.

eolivelli commented:
@amitvc great.

I am going to cut a release during the weekend or on Monday.
If you perform some smoke tests it will be great.
Ping me if you have problems

amitvc commented:
@eolivelli I will do some testing locally first thing Monday morning.
